lib/Doctrine/DBAL/Query/QueryBuilder.php

```diff
@@ -997,76 +1084,55 @@ public function orderBy(string $sort, ?string $order = null) : self
      */
     public function addOrderBy(string $sort, ?string $order = null) : self
     {
-        return $this->add('orderBy', $sort . ' ' . (! $order ? 'ASC' : $order), true);
-    }
+        $orderBy = $sort;
 
-    /**
-     * Gets a query part by its name.
-     *
-     * @return mixed
-     */
-    public function getQueryPart(string $queryPartName)
-    {
-        return $this->sqlParts[$queryPartName];
-    }
+        if ($order !== null) {
+            $orderBy .= ' ' . $order;
+        }
 
-    /**
-     * Gets all query parts.
-     *
-     * @return array<string, mixed>
-     */
-    public function getQueryParts() : array
-    {
-        return $this->sqlParts;
+        $this->orderBy[] = $orderBy;
+
+        $this->state = self::STATE_DIRTY;
+
+        return $this;
     }
 
     /**
-     * Resets SQL parts.
-     *
-     * @param array<int, string>|null $queryPartNames
-     *
-     * @return $this This QueryBuilder instance.
+     * @throws QueryException
      */
-    public function resetQueryParts(?array $queryPartNames = null) : self
+    private function getSQLForSelect() : string
     {
-        if ($queryPartNames === null) {
-            $queryPartNames = array_keys($this->sqlParts);
+        if (! $this->select) {
+            throw new QueryException('No SELECT expressions given. Please use select() or addSelect().');
         }
 
-        foreach ($queryPartNames as $queryPartName) {
-            $this->resetQueryPart($queryPartName);
+        $query = 'SELECT';
+
+        if ($this->distinct) {
+            $query .= ' DISTINCT';
         }
 
-        return $this;
-    }
+        $query .= ' ' . implode(', ', $this->select);
 
-    /**
-     * Resets a single SQL part.
-     *
-     * @return $this This QueryBuilder instance.
-     */
-    public function resetQueryPart(string $queryPartName) : self
-    {
-        $this->sqlParts[$queryPartName] = self::SQL_PARTS_DEFAULTS[$queryPartName];
+        if ($this->from) {
+            $query .= ' FROM ' . implode(', ', $this->getFromClauses());
+        }
 
-        $this->state = self::STATE_DIRTY;
+        if ($this->where) {
```

Comment:
Sorry for the last minute request, but these [issues](https://scrutinizer-ci.com/g/doctrine/dbal/inspections/4ab891ff-68b0-4ff6-a2cd-d5d7833a9c7d/issues/) found by Scrutinizer seem worth addressing. Especially this one.

Since `$this->where` is `string|ComplexExpression|null`, it can be assigned a string `'0'` and ignored. There should be an `if ($this->where !== null)` comparison.

Reply:
Good point. This was `CompositeExpression|null` before, so implicit conversion to boolean was OK. Now that's it can be a `string`, it's indeed become a potential issue.

Fixed.